arch/riscv:initial support for floating-point

no1wudi opened the pull request.

Commit:
arch/riscv: Initial support for floating-point

Add floating-register load/store for riscv

Signed-off-by: Huang Qi <757509347@qq.com>

ioannisg commented:
Just a note: please correct the 2nd commit message, according to the guidelines.
Thanks for your contribution!

nategraff-sifive reviewed (changes requested):
Thanks for this work!

So the modular nature of RISC-V means that XLEN, the number of bits in the general-purpose `x*` registers is independent of FLEN, the number of bits in the floating point registers. For example, `rv32imfd` and `rv64imf` are two valid ISA strings where XLEN != FLEN.

In these patches, `RV_OP_LOADFPREG` and `RV_OP_STOREFPREG` definitions depend on `CONFIG_64BIT`, which for RISC-V in Zephyr should probably just describe XLEN. Maybe we don't have a good way to describe this in Zephyr, but what we really need here is a `CONFIG_FLOAT_64BIT` option.

Commit:
arch: riscv: Add CONFIG_FLOAT_64BIT option

Use CONFIG_FLOAT_64BIT to determine fp register width,
It seems that only riscv need to know the width of fp registers

Signed-off-by: Huang Qi <757509347@qq.com>

Commit:
arch: riscv: Add floating point document

Add simple float desciption for riscv

Signed-off-by: Huang Qi <757509347@qq.com>

Commit:
arch: riscv: Fix code style

Fix code style in comments befor float point register define in context

Signed-off-by: Huang Qi <757509347@qq.com>

nategraff-sifive reviewed (changes requested):
Thanks for updating your PR!

I had another thought: the floating point extensions define some CSRs/CSR fields, which should probably be initialized to the proper values for floating point operation.

I'm not sure this is a complete list, but you might want to start with:
- [RISC-V Unprivileged ISA](https://riscv.org/specifications/) Section 11.2 Floating Point Control and Status Register
- [RISC-V Privileged ISA](https://riscv.org/specifications/privileged-isa/) Section 3.1.6.5 Extension Context Status in `mstatus` Register

Probably a good idea to make sure that floating point is turned on and has the desired rounding mode, as well as clearing any outstanding exceptions in case any of these things aren't initialized or cleared by the implementation on cold boot or reset.

Commit:
arch: riscv: Init float control register after boot & reset

Clear all exception flags,and set float rounding mode to
IEEE-754 default.

Signed-off-by: Huang Qi <757509347@qq.com>

no1wudi commented:
> Thanks for updating your PR!
> 
> I had another thought: the floating point extensions define some CSRs/CSR fields, which should probably be initialized to the proper values for floating point operation.
> 
> I'm not sure this is a complete list, but you might want to start with:
> 
>     * [RISC-V Unprivileged ISA](https://riscv.org/specifications/) Section 11.2 Floating Point Control and Status Register
> 
>     * [RISC-V Privileged ISA](https://riscv.org/specifications/privileged-isa/) Section 3.1.6.5 Extension Context Status in `mstatus` Register
> 
> 
> Probably a good idea to make sure that floating point is turned on and has the desired rounding mode, as well as clearing any outstanding exceptions in case any of these things aren't initialized or cleared by the implementation on cold boot or reset.

Thanks for reviewing, and there seems to be another question.
 
Should we store a fcsr(fflags) for each thread? 

RISC-V do not support traps on floating-point exceptions in the base ISA, and require explicit checks of the flags in software.

nategraff-sifive commented:
> Should we store a fcsr(fflags) for each thread?

We should save the entire fcsr in the thread context.

palmer-dabbelt reviewed (approved):
I'm not a Zephyr developer, so I can't tell if this is complete, but nothing here appears to be incorrect to me.

Thanks!

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,31 @@ when there is no need to do so.
+RISCV architecture
+------------------
+
+On the RISCV architecture, the kernel treats each thread as a non-user
```
For clarity, how about:
```
On the RISCV architecture, the kernel treats each thread as
a non-FPU user unless the thread is tagged as an FPU user
using one of the following techniques: 
```

What are the consequences if a thread does floating point operations
but was not tagged as an FPU user?  Context switches won't be
saving/restoring the FP registers so bad things will happen, right?
If so, this should be emphasized here.

no1wudi replied:
Thanks for reviewing, I will update the document later according to your suggestion.

dbkinder replied:
Oh, so the default for RISCV is all threads are treated as FPU users?  The docs should certainly say that.

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,31 @@ when there is no need to do so.
+  :c:macro:`K_FP_REGS` to :cpp:func:`k_thread_create()`.
+
+During thread context switching the RISCV kernel saves the *callee-saved*
+floating point registers, if the switched-out thread has been using them.
```
should this read ``if the switched-out thread is tagged as an FPU user``?

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,31 @@ when there is no need to do so.
+During thread context switching the RISCV kernel saves the *callee-saved*
+floating point registers, if the switched-out thread has been using them.
+Additionally, the *caller-saved* floating point registers are saved on
+the thread's stack. If the switched-in thread has been using the floating
```
same here, should this read ``If the switched-in thread is tagged as an FPU-user, the kernel...``

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,31 @@ when there is no need to do so.
+the thread's stack. If the switched-in thread has been using the floating
+point registers, the kernel restores the *callee-saved* FP registers of
+the switched-in thread and the *caller-saved* FP context is restored from
+the thread's stack. Thus, the kernel does not save or restore the FP
```
Change ``Thus, the kernel`` to ``The kernel``

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,31 @@ when there is no need to do so.
+point registers, the kernel restores the *callee-saved* FP registers of
+the switched-in thread and the *caller-saved* FP context is restored from
+the thread's stack. Thus, the kernel does not save or restore the FP
+context of threads that are not using the FP registers. An extra 128 bytes
```
should this read, ``threads that are not tagged as an FPU-user``?

Commit:
arch: riscv: Add fcsr in thread fp context

Add 32bit width fcsr in thread fp context and update docs.

Signed-off-by: Huang Qi <757509347@qq.com>

Commit:
arch: riscv: Fix typo in fp docs

Correct CONFIG_FP to CONFIG_FLOAT in float point services docs.

Signed-off-by: Huang Qi <757509347@qq.com>

Commit:
arch: riscv: Remove trailing whitespace

Remove trailing whitespace in docs for floating point service

Signed-off-by: Huang Qi <757509347@qq.com>

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -193,6 +193,19 @@ when there is no need to do so.
+RISCV architecture
+------------------
+
+On the RISCV architecture, the kernel simply treats all thread as an FPU user
```
Well, that certainly simplifies things.
How about this to fix some misspellings and for clarity:
```
On the RISCV architecture, the kernel treats all threads as an FPU user
if :option:`CONFIG_FLOAT` and :option:`CONFIG_FP_SHARING` are
selected. This may lead to performance problems because the FP
registers will be saved and restored on every thread context switch.

In an optimized implementation, the kernel should only save and restore
FP registers on a thread context switch if the thread is tagged as an FPU
user. An extra 132 bytes (4 x 32 single precision + fscr)
or 260 bytes (8 x 32 double precision + fscr) of stack space is
required to load and store floating point registers.  
```

nategraff-sifive reviewed (approved):
Code LGTM, thanks for your contribution!

Commit:
arch: riscv: Update docs

Update docs for floating point services.

Signed-off-by: Huang Qi <757509347@qq.com>

dbkinder reviewed (approved):
+1 for docs, hopefully, the "optimized implementation" is coming (and we'd need to update the docs then)

nashif reviewed (changes requested):
You need to enable float tests for riscv, nothing right now is testing your changes.

nategraff-sifive commented:
> You need to enable float tests for riscv, nothing right now is testing your changes.

Probably the best bet here is to use the `sifive_u` target in QEMU's RISC-V support, since it supports the `rv32imafd` and `rv64imafd` ISAs.

no1wudi commented:
> > You need to enable float tests for riscv, nothing right now is testing your changes.
> 
> Probably the best bet here is to use the `sifive_u` target in QEMU's RISC-V support, since it supports the `rv32imafd` and `rv64imafd` ISAs.

Which board can run with this target? I tried it with qemu_riscv64 by "-machine sifive_u", but got:
 qemu-system-riscv64: warning: No -bios option specified. Not loading a firmware.
qemu-system-riscv64: warning: This default will change in a future QEMU release. Please use the -bios option to avoid breakages when this happens.
qemu-system-riscv64: warning: See QEMU's deprecation documentation for details.
qemu-system-riscv64: warning: nic cadence_gem.0 has no peer

Commit:
arch: riscv: Add float point tests implementions

Add float context and sotre/load float registers implemention
for tests/kernel/fp_sharing/generic

Signed-off-by: Huang Qi <757509347@qq.com>

nategraff-sifive commented:
The sifive_u target needs its own Zephyr "board" because it's a fundamentally different target from the current QEMU RISC-V simulator (which uses the `sifive_e` machine) in Zephyr. Luckily QEMU is happy to spit a devicetree out for you if you also pass `-machine dumpdtb=sifive_u.dtb`

Also the bios warning you're seeing isn't an issue. You can disable that one by passing `-machine sifive_u,firmware=none`

tyrantlizard77 reviewed (commented):
I agree with @nashif, we really need this enabled for an emulated board target to demonstrate that this actually works and keep it from bit-rotting.

I'm curious why you didn't implement lazy fp save/restore, as we have done on other floating point supporting targets? Is this a planned enhancement for later or is there something about this arch which prevents it from being implemented for Zephyr?

no1wudi commented:
> I agree with @nashif, we really need this enabled for an emulated board target to demonstrate that this actually works and keep it from bit-rotting.
> 
> I'm curious why you didn't implement lazy fp save/restore, as we have done on other floating point supporting targets? Is this a planned enhancement for later or is there something about this arch which prevents it from being implemented for Zephyr?

As the titile says, this PR only implements the "initial" support for floating support.
I'm porting it to a new platform that supports floating-point instructions, which is still in its early stages, so the code related to the arch was submited first. I plan to implement the basic functions before considering further optimization.
I tested the code on my board, But I‘m not familiar with QMEU, maybe Maybe this PR needs to wait for the relevant emulated board target to be added.

andrewboie commented:
Not opposed to iterative refinements, but test coverage is hard requirement, need something to prove that what we have so far works.

xodus7 commented:
@no1wudi I'm wondering what is the current state of this work?
